python/ray/air/config.py

```diff
@@ -307,15 +307,15 @@ class DatasetConfig:
         transform: Whether to transform the dataset with the fitted preprocessor.
             This must be enabled at least for the dataset that is fit.
             True by default.
-        use_stream_api: Whether the dataset should be streamed into memory using
-            pipelined reads. When enabled, get_dataset_shard() returns DatasetPipeline
-            instead of Dataset. The amount of memory to use is controlled
-            by `stream_window_size`. False by default.
-        stream_window_size: Configure the streaming window size in bytes.
-            A good value is something like 20% of object store memory.
-            If set to -1, then an infinite window size will be used (similar to
-            bulk ingest). This only has an effect if use_stream_api is set.
-            Set to 1.0 GiB by default.
+        max_object_store_memory_fraction [Experimental]: The maximum fraction
+            of Ray's shared-memory object store to use for the transformed
+            Dataset. The default value is -1, meaning that the preprocessed
+            dataset should be cached, which may cause spilling if its size is
+            larger than the object store's capacity. Pipelined ingest (all
+            other values, from 0 to 1) is experimental. Note that the absolute
+            memory capacity used is based on the object store capacity at
+            invocation time; this does not currently cover autoscaling cases
+            where the size of the cluster may change.
```

Comment:
I think we should explicitly clarify here and in the check-ingest docs that the object store usage uses the initial dataset size, **before preprocessing**.

Also that it's a **per-window** fraction, and that there can be multiple windows at a time since it's pipelined.

Reply:
It is pretty hard to reason about the object store size, especially since it's unclear to a user exactly how large the preprocessed dataset will be.

Ultimately, we expect the dataset optimizer to automatically determine the best configuration for this right? If not, then we may want to consider moving to a row-based API instead of a bvtes-based API in the future.

Reply:
I don't think we should worry about this too much. The current DatasetPipeline version just requires too much knowledge of implementation internals that I think it will actually be more confusing to users...I would rather just leave this experimental for now and we can revisit the docs once the implementation is more stable. Does that sound OK?

> It is pretty hard to reason about the object store size, especially since it's unclear to a user exactly how large the preprocessed dataset will be.
> 
> Ultimately, we expect the dataset optimizer to automatically determine the best configuration for this right? If not, then we may want to consider moving to a row-based API instead of a bvtes-based API in the future.

The idea behind exposing the object store size (instead of fraction of dataset size) is:
1. it does not depend on what dataset is used, so it does not have to be changed based on which dataset is used
2. it's more directly related to what the user wants (e.g., prevent spilling vs. prioritize caching)
3. it's more compatible with autoscaling
The end state would be to set this to a good default like 50-80% and most users will never have to change it.